Feature/benchmark resnet18

LiamHe opened the pull request:
This Pull Request

1. Parse options for each class separately;
2. Implements a resnet 18 example in DAWNBench (bkj). (Using MultiStepLR schedule in ResNet's paper instead of `linear_cycle` in theirs.)
3. Refactor some of other modules.

martinjaggi commented:
thanks guys, this looks very nice!

Panaetius commented:
I just tested it and it works good. Though the dashboard runs into a timeout exception and loses connection. 

We should set `_request_timeout` to `None`  in https://github.com/mlbench/mlbench/blob/999c01da88f9ca01010b4b79ea799eed99b99df0/mlbench/master/api/utils/run_utils.py#L70 to prevent this

LiamHe commented on mlbench/master/api/utils/run_utils.py:
```diff
@@ -54,7 +54,9 @@ def run_model_job(model_run, experiment="test_MPI"):
             '--run_id',
-            model_run.id]
+            model_run.id,
+            '--config-file', '/codes/configs/debug_cifar10.json'
```
You are right. I use the config files because applying changes to `run_model_job` seems not the same as others. Otherwise adding config files from dashboard/command line would be better.
The function body of `run_model_job` remain the same even after `helm upgrade`. Do you know if there are any ways to apply changes to `run_utils.py`-`run_model_job` easily?

LiamHe commented on mlbench/refimpls/pytorch/controlflow/controlflow.py:
```diff
@@ -0,0 +1,170 @@
+    # all_reduce the gradients.
+    for ind, param in enumerate(model.parameters()):
+        # all reduce.
+        dist.all_reduce(param.grad.data, op=dist.reduce_op.SUM)
```
This module seems to only work for `nccl` and `gloo` backend. But other modules in `pytorch/torch/nn/parallel` could be useful for mpi. We can check what could be beneficial.

LiamHe commented:
@IamTao The original models `6n+2` are also implemented. The ResNet-18 here is from an implementation from DAWNBench which achieves 2rd fastest in terms of training time.

IamTao commented:
@LiamHe <https://github.com/LiamHe>, I see. maybe we should add a few
comments and also point to their repository (they said they made minor
modifications). Also, is it a correct way to do so? i.e., directly include
non-official implementation (but improved variant) to our official
reference implementation. It is always possible to have improved
implementation in the future. Just a doubt.


On Mon, Aug 27, 2018 at 11:27 AM Lie He <notifications@github.com> wrote:

> @IamTao <https://github.com/IamTao> The original models 6n+2 are also
> implemented. The ResNet-18 here is from an implementation from DAWNBench
> which achieves 2rd fastest in terms of training time.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mlbench/mlbench/pull/45#issuecomment-416169552>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHcbf323IerEkWbIkWTdmU1Nh4ANXC_3ks5uU7uDgaJpZM4WJCj2>
> .
>

LiamHe commented:
@IamTao Sure, I have added a comment in `feature/benchmark-resnet18` branch, it has not been merge yet. 
I added them because they have provided an epoch-time-top1Accuracy file on V100 instance. By running same model on difference devices, we can know if we have introduces other performance bottlenecks. Since they seems to be adapted from `post-activation` version of ResNet, we can replace resnet-18 with the standard one and move it to some `contrib/` directories/repositories.

IamTao commented:
@LiamHe <https://github.com/LiamHe> the idea of
contrib/directories/repositories
sounds good. Also, since our current implementation is based on
`pre-activation` (same as `torchvision`), do you think we should also add
the option to use the block of `post-activation`?

On Mon, Aug 27, 2018 at 11:58 AM Lie He <notifications@github.com> wrote:

> @IamTao <https://github.com/IamTao> Sure, I have added a comment in
> feature/benchmark-resnet18 branch, it has not been merge yet.
> I added them because they have provided an epoch-time-top1Accuracy file on
> V100 instance. By running same model on difference devices, we can know if
> we have introduces other performance bottlenecks. Since they seems to be
> adapted from post-activation version of ResNet, we can replace resnet-18
> with the standard one and move it to some contrib/
> directories/repositories.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mlbench/mlbench/pull/45#issuecomment-416177449>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHcbf_FO9TSEzsKKWDPbL7KWMBxCUP1nks5uU8K_gaJpZM4WJCj2>
> .
>
